internal/controllers/virtual_devices_controller.go

```diff
@@ -0,0 +1,210 @@
+
+	uTokenID, err := strconv.ParseUint(tokenID, 10, 64)
+	if err != nil {
+		return fiber.NewError(fiber.StatusBadRequest, "invalid tokenID provided")
+	}
+
+	user, err := vc.usersClient.GetUser(c.Context(), &pb.GetUserRequest{
+		Id: userID,
+	})
+	if err != nil {
+		vc.log.Debug().AnErr("GetUser Grpc", err).Msg("error occurred when fetching user")
+		return fiber.NewError(fiber.StatusUnauthorized, "error occurred when fetching user")
+	}
+
+	if user.EthereumAddress == nil {
+		return fiber.NewError(fiber.StatusUnauthorized, "User does not have an Ethereum address on file.")
+	}
+
+	integration, err := vc.deviceDefSvc.GetIntegrationByTokenID(c.Context(), uTokenID)
+	if err != nil {
+		return fiber.NewError(fiber.StatusBadRequest, "failed to get integration")
+	}
+
+	vid, err := strconv.ParseInt(vehicleNode, 10, 64)
+	if err != nil {
+		return fiber.NewError(fiber.StatusBadRequest, "invalid vehicleNode provided")
+	}
+	vnID := types.NewNullDecimal(decimal.New(vid, 0))
+	vehicleNFT, err := models.VehicleNFTS(
+		models.VehicleNFTWhere.TokenID.EQ(vnID),
+		models.VehicleNFTWhere.OwnerAddress.EQ(null.BytesFrom(common.HexToAddress(*user.EthereumAddress).Bytes())),
+	).Exists(c.Context(), vc.DBS().Reader)
+	if err != nil {
+		vc.log.Error().Err(err).Str("vehicleNode", vehicleNode).Str("tokenID", tokenID).Msg("Could not fetch minting payload for device")
+		return fiber.NewError(fiber.StatusInternalServerError, "error generating device mint payload")
+	}
+
+	if !vehicleNFT {
+		return fiber.NewError(fiber.StatusBadRequest, "user does not own vehicle node")
+	}
+
+	response := vc.getVirtualDeviceMintPayload(int64(integration.TokenId), vid)
+
+	return c.JSON(response)
+}
+
+// SignVirtualDeviceMintingPayload godoc
+// @Description validate signed signature for vehicle minting
+// @Tags        integrations
+// @Produce     json
+// @Success     200 {array}
+// @Router      /integration/:tokenID/mint-virtual-device [post]
+func (vc *VirtualDeviceController) SignVirtualDeviceMintingPayload(c *fiber.Ctx) error {
+	tokenID := c.Params("tokenID")
+	vehicleNode := c.Query("vehicle_id")
+	userID := helpers.GetUserID(c)
+
+	req := &SignVirtualDeviceMintingPayloadRequest{}
+	if err := c.BodyParser(req); err != nil {
+		return fiber.NewError(fiber.StatusBadRequest, "Couldn't parse request.")
+	}
+
+	signature, err := hex.DecodeString(req.OwnerSignature)
+	if err != nil {
+		return fiber.NewError(fiber.StatusBadRequest, "invalid signature provided")
+	}
+
+	if len(signature) != 65 {
+		return fiber.NewError(fiber.StatusBadRequest, "invalid signature provided")
+	}
+
+	uTokenID, err := strconv.ParseUint(tokenID, 10, 64)
+	if err != nil {
+		return fiber.NewError(fiber.StatusBadRequest, "invalid tokenID provided")
+	}
+
+	integration, err := vc.deviceDefSvc.GetIntegrationByTokenID(c.Context(), uTokenID)
+	if err != nil {
+		return helpers.GrpcErrorToFiber(err, "failed to get integration")
+	}
+
+	user, err := vc.usersClient.GetUser(c.Context(), &pb.GetUserRequest{
```

Comment:
Do we check that this person owns the NFT? That it exists?

I think we could consolidate the following logic:

1. User on file
2. User has address
3. Vehicle NFT exists
4. User owns NFT

I'm sure it's annoying to write twice.

Reply:
Let's store `userAddr = common.HexToAddress(*user.EthereumAddress)`.